src/cargo/core/registry.rs

```diff
@@ -274,39 +314,95 @@ impl<'cfg> Registry for PackageRegistry<'cfg> {
     fn query(&mut self,
              dep: &Dependency,
              f: &mut FnMut(Summary)) -> CargoResult<()> {
-        // Ensure the requested source_id is loaded
-        self.ensure_loaded(dep.source_id(), Kind::Normal).chain_err(|| {
-            format!("failed to load source for a dependency \
-                     on `{}`", dep.name())
-        })?;
-
-
         let (override_summary, n, to_warn) = {
             // Look for an override and get ready to query the real source.
             let override_summary = self.query_overrides(&dep)?;
-            let source = self.sources.get_mut(dep.source_id());
-            match (override_summary, source) {
-                (Some(_), None) => bail!("override found but no real ones"),
-                (None, None) => return Ok(()),
-
-                // If we don't have an override then we just ship everything
-                // upstairs after locking the summary
-                (None, Some(source)) => {
-                    let locked = &self.locked;
-                    return source.query(dep, &mut |summary| f(lock(locked, summary)))
+
+            // Next up on our list of candidates is to check the `[augment]`
+            // section of the manifest. Here we look through all augmentations
+            // relevant to the source that `dep` points to, and then we match
+            // name/version. Note that we don't use `dep.matches(..)` because
+            // the augmentations, by definition, come from a different source.
+            // This means that `dep.matches(..)` will always return false, when
+            // what we really care about is the name/version match.
```

Comment:
Yeah old overrides are... odd...

In any case though the suggestion sounds good!